icu4c/source/i18n/listformatter.cpp

```diff
@@ -474,42 +473,43 @@ UnicodeString& ListFormatter::format_(
                 &offsetFirst,
                 &offsetSecond,
                 errorCode);
-        if (offsets != nullptr) {
-            offsets[nItems - 1] = offsetSecond;
-            // Adjust the offset since the joinStringsAndReplace may shift
-            // the starting point of the first item depending on the pattern.
-            for (int32_t j = 0; j < nItems - 1; ++j) {
-                offsets[j] += offsetFirst;
-            }
+        if (handler != nullptr) {
+            prefixLength += offsetFirst;
+            offsets[nItems - 1] = offsetSecond - prefixLength;
         }
     }
-    if (offsets != nullptr && handler != nullptr) {
+    if (handler != nullptr) {
         // If there are already some data in appendTo, we need to adjust the index
         // by shifting that lenght while insert into handler.
-        int32_t shift = appendTo.length();
-        int32_t lastAdded = 0;
+        int32_t shift = appendTo.length() + prefixLength;
+        // Output the ULISTFMT_ELEMENT_FIELD in the order of the input elements
         for (int32_t i = 0; i < nItems; ++i) {
-            if (offsets[i] > lastAdded) {
-                handler->addAttribute(
-                    ULISTFMT_LITERAL_FIELD,  // id
-                    shift + lastAdded,  // index
-                    shift + offsets[i]);  // limit
-            }
-            lastAdded = offsets[i] + items[i].length();
+            offsets[i + nItems] = offsets[i] + items[i].length() + shift;
+            offsets[i] += shift;
             handler->addAttribute(
                 ULISTFMT_ELEMENT_FIELD,  // id
-                shift + offsets[i],  // index
-                shift + lastAdded);  // limit
+                offsets[i],  // index
+                offsets[i + nItems]);  // limit
         }
-        // If there are text after the last itme, we should also insert a literal.
-        if (result.length() > lastAdded) {
+        // The locale pattern may reorder the items (such as in ur-IN locale),
+        // so we cannot assume the array is in accendning order.
+        // To handle the edging case, just insert the two ends into the array
+        // and sort. Then we output ULISTFMT_LITERAL_FIELD if the indecies
+        // between the even and odd position are not the same in the sorted array.
+        offsets[2 * nItems] = shift - prefixLength;
+        offsets[2 * nItems + 1] = result.length() + shift - prefixLength;
```

Comment:
If I'm following correctly, the indices before sorting are as follows:

```
blah blah |prefix|apples|, |oranges|, and |cherries|suffix|
          6      0      3  1       4      2        5      7
```

Seems to be right.

Reply:
yes